manim/mobject/mobject.py

```diff
@@ -1295,3 +1306,20 @@ class Group(Mobject):
     def __init__(self, *mobjects, **kwargs):
         Mobject.__init__(self, **kwargs)
         self.add(*mobjects)
+
+
+class _AnimationBuilder:
+    def __init__(self, mobject):
+        self.mobject = mobject
+
+    def __call__(self, *method_args, **method_kwargs):
+        self.method_args = method_args
+        self.method_kwargs = method_kwargs
+        return self
+
+    def __getattr__(self, method_name):
+        self.method_name = method_name
+        return self
+
+    def animation_data(self):
+        return self.mobject, self.method_name, self.method_args, self.method_kwargs
```

Comment:
Is there a benefit to not having this return a function in `__getattr__` that returns an animation? That would simplify the logic in `Scene.play` even more, wouldn't it? And then you wouldn't have to make the `animate` attribute a property because you could just keep the same object around because that object wouldn't be holding any state anymore.

I'm also not especially a fan of the fact that you could do `obj.animate(arg, kwarg=kwarg).method`, or even `obj.animate.method1.method2(arg, kwarg=kwarg)`, or you could even chain calls too. Those are things that are unlikely to occur, but they seem to me be suboptimal things that come about from the actual object holding state beyond the mobject it's animating.

Reply:
Also with the current system, you'd get an `AttributeError` when `self.method_name` is used to get the actual method, somewhere within `Scene.play`, instead of right away when you try to access the method. That could be changed by checking `hasattr` or storing the method instead of the method name, but that feels more like a duct tape fix than a fulfilling solution

Reply:
If you did do it with a function it would have to have a closure over the mobject and the string used to initialize the function, so it'd end up looking a lot like the way it does now.

The point about chaining does seem strange to me, so I can add a check preventing recalling `__getattr__` when the data is already set.

I'm not sure what you mean about the `AttributeError` since `self.method_name` is already used to get the method here: https://github.com/ManimCommunity/manim/pull/881/files#diff-b2a9bbdd948ff6b81ec07f5d3c39a716cbd14d16d240cc05458d482232ba8e51R637. If you mean that `__getattr__` should check that the mobject has attribute I can add that check, but then there's the check of if the attribute is a method and if it's a method for animating, which I don't think can be checked well.

Reply:
> If you did do it with a function it would have to have a closure over the mobject and the string used to initialize the function, so it'd end up looking a lot like the way it does now.

I think it would make the code at least better organized and perhaps more readable if `__getattr__` returned a function that returns an animation. Instead of getting the method all the way in some other file and turning it into an animation there, you'd do it in the same place that this class is defined. Additionally, if the function `__getattr__` returns returns an animation, then `Scene.play` only has to worry about animations, nothing else, not even these `_AnimationBuilder` objects.

> If you mean that `__getattr__` should check that the mobject has attribute I can add that check, but then there's the check of if the attribute is a method and if it's a method for animating, which I don't think can be checked well.

I do mean this. I care because it'll lead to less obscure errors to the user if they e.g. make a typo, so that the `AttributeError` would be raised when *they* access the attribute instead of once the library does in a more obscure place. And whether the attribute is a method can be checked, with `inspect.ismethod` and I think there might already be a function in manim for checking if something is a mobject method. But as to whether it can be animated, I'm not sure how well that can be checked, if at all. But I mainly care about the error for accessing the method *at all*, not really the other stuff.

And I suppose another benefit of making of `__getattr__` return a function that returns an animation is that whether your arguments are invalid would be checked when you call the returned function, rather than somewhere in `Scene.play`, yet again leading to less obscure errors.

Reply:
I see what you mean now and I did it, although it's different behavior from the previous code now because the animation is initialized when the arguments are passed rather than when `Scene.play()` is called.

It also exposes some problems with our imports, because animation.py can't be imported from mobject.py at the top of the file. Some refactoring is probably in order.

Reply:
One small thing is that the .animate attribute doesn't need to be a property anymore because it doesn't need to be regenerated each time it's accessed, but it doesn't really matter much

Reply:
It actually causes an error: `'function' object has no attribute 'shift'`.

Reply:
I think you may have misunderstood me, I meant to add `self.animate = _AnimationBuilder(self)` to Mobject's `__init__`, not to just remove the `@property` decorator

Reply:
It turns out it causes infinite recursion when `Mobject.generate_target()` is called since `Mobject` has a reference to a `_AnimationBuilder` and the `_AnimationBuilder` has a reference to the `Mobject`, although I wasn't trying to avoid that at all.

Reply:
Ahh ok, fair enough then.